otter/util/zk.py

```diff
@@ -170,3 +199,113 @@ def get_zk_dispatcher(kz_client):
             partial(perform_get_stat, kz_client),
         AcquireLock: perform_acquire_lock
     })
+
+
+class PollingLock(object):
+    """
+    Zookeeper lock recipe that polls the children on interval basis instead
+    of leaving a watch on previous child. It's supposed to replace
+    :obj:`kazoo.recipe.lock.Lock` and has same signature and _similar_ behavior
+    for ``acquire`` and ``release`` methods.
+    """
+
+    def __init__(self, dispatcher, path, identifier=""):
+        self.dispatcher = dispatcher
+        self.path = path
+        self.identifier = identifier
+        self._node = None
```

Comment:
> Side effect? Sigh, defect :).

This is bare minimal internal state required to function. Please compare with [kazoo](https://github.com/python-zk/kazoo/blob/master/kazoo/recipe/lock.py#L74-L107) :)

> This is a mostly beautiful and apparently correct implementation of the locking logic

Thanks!

> which I think I understand

https://zookeeper.apache.org/doc/trunk/recipes.html#sc_recipes_Locks without the watch part.

> Except for this piece. _node is set when the lock is going to be acquired…

I am sorry this is confusing. I am assuming you mean "node is released when locking is going to be acquired". This is done to allow leader election behavior (long-held lock) being used in selfheal service along with immediate acquire/release lock behavior used when modifying state. The idea is one worker will acquire the lock and check if it remains acquired by calling `is_lock_acquired` at some interval. If it is not acquired which can happen if the session was re-established it will call `acquire` again. Now, what should the function do if it already finds `_node` in such a situation? I believe the sanest thing to do is delete it (by calling `release_eff`) before beginning to acquire the lock. This way that child node (if it still remains for whatever reason) is not hanging around. Child node remaining is the original problem this code is trying to solve. I'll add docs about this like you suggested. 

> And then a bunch of tests set it directly in order to test things about that internal state, rather than driving it through the external interface.

I could change the tests to drive through external interface but I believe that is being tested in `test_acquire_*` functions. I am trying to think if it is a bad idea to test internal state without driving it from public interface. Not sure.

> I struggled to think of how to deploy Active Nothing here, although I'm sure if we could figure it out the code would be the better for it.

Hmm. I don't know how that pattern will help here. Moreover, it looks you want to use it because you want to use it :) ? Did you have any reasons for this?